### PR TITLE
Add support for for selection of metadata to -m

### PR DESCRIPTION
**Pull request by tovrstra**
This fixes at least a part of #98. (See issue for more details. The `-m` option has been extended, `-M` is not included.)

@psychemedia Could you try if this also works for you? Thanks! I've also updated the `README.md` with some practical instructions.

### CODE REVIEW

File: src/nb_clean/__init__.py [hunk 1]
```diff
@@ -86,8 +87,10 @@ def add_git_filter(
     ----------
     remove_empty_cells : bool, default False
         If True, remove empty cells.
-    preserve_cell_metadata : bool, default False
-        If True, preserve cell metadata.
+    preserve_cell_metadata : list of str or None, default None
+        If None, clean all cell metadata.
+        If [], preserve cell metadata.
```
Comment (yasirroni) [inline src/nb_clean/__init__.py:92]: What does this imply?

Comment (yasirroni) [reply src/nb_clean/__init__.py:92]: Now I understand that `[]` used to behave when `--preserve-cell-metadata` is given without any arguments.

Comment (yasirroni) [reply src/nb_clean/__init__.py:92]: Maybe edit to clearly state "all cell metadata"?

Comment (tovrstra) [reply src/nb_clean/__init__.py:92]: Agreed, that would be better. I'll also try to clarify the reason `[]` has this behavior.

File: tests/test_check_notebook.py [hunk 2]
```diff
@@ -37,18 +39,87 @@ def test_check_notebook_remove_empty_cells(
...
+    assert nb_clean.check_notebook(
+        clean_notebook_with_metadata,
+        preserve_cell_metadata=preserve_cell_metadata,
+    ) is (
+        (preserve_cell_metadata is not None)
+        and (
+            preserve_cell_metadata == []
+            or {"tags", "special", "nbclean"}.issubset(preserve_cell_metadata)
+        )
+    )
```
Comment (srstevenson) [inline tests/test_check_notebook.py:67]: Given this is quite a long assertion now I think it'd be worth refactoring out `output` and `expected` variables:
```python
expected = (preserve_cell_metadata is not None...
output = nb_clean.check_notebook...
assert output is expected
```

File: tests/test_check_notebook.py [hunk 3]
```diff
@@ -37,18 +39,87 @@ def test_check_notebook_remove_empty_cells(
...
+    assert nb_clean.check_notebook(
+        clean_notebook_with_tags_metadata,
+        preserve_cell_metadata=preserve_cell_metadata,
+    ) is (
+        (preserve_cell_metadata is not None)
+        and (
+            preserve_cell_metadata == []
+            or {"tags"}.issubset(preserve_cell_metadata)
+        )
+    )
```
Comment (srstevenson) [inline tests/test_check_notebook.py:95]: Ditto.

File: tests/test_check_notebook.py [hunk 4]
```diff
@@ -37,18 +39,87 @@ def test_check_notebook_remove_empty_cells(
...
+    """Test nb_clean.check_notebook when preserving cell metadata."""
+    assert nb_clean.check_notebook(
+        clean_notebook_with_tags_special_metadata,
+        preserve_cell_metadata=preserve_cell_metadata,
+    ) is (
+        (preserve_cell_metadata is not None)
+        and (
+            preserve_cell_metadata == []
+            or {"tags", "special"}.issubset(preserve_cell_metadata)
         )
```
Comment (srstevenson) [inline tests/test_check_notebook.py:122]: Ditto.

File: tests/test_clean_notebook.py [hunk 5]
```diff
@@ -36,17 +37,63 @@ def test_clean_notebook_remove_empty_cells(
...
+    [
+        [],
+        ["nbclean", "tags", "special"],
+        ["nbclean", "tags", "special", "toomany"],
+    ],
+)
 def test_clean_notebook_preserve_metadata(
     dirty_notebook: nbformat.NotebookNode,
     clean_notebook_with_metadata: nbformat.NotebookNode,
+    preserve_cell_metadata,
```
Comment (srstevenson) [inline tests/test_clean_notebook.py:51]: Let's add a type annotation to this.

File: tests/test_clean_notebook.py [hunk 6]
```diff
@@ -36,17 +37,63 @@ def test_clean_notebook_remove_empty_cells(
...
     )
 
 
+@pytest.mark.parametrize(
+    "preserve_cell_metadata", [["tags"], ["tags", "toomany"]]
+)
+def test_clean_notebook_preserve_metadata_tags(
+    dirty_notebook: nbformat.NotebookNode,
+    clean_notebook_with_tags_metadata: nbformat.NotebookNode,
+    preserve_cell_metadata,
```
Comment (srstevenson) [inline tests/test_clean_notebook.py:68]: Ditto on type annotation.

File: tests/test_clean_notebook.py [hunk 7]
```diff
@@ -36,17 +37,63 @@ def test_clean_notebook_remove_empty_cells(
...
+
+
+@pytest.mark.parametrize(
+    "preserve_cell_metadata",
+    [["tags", "special"], ["tags", "special", "toomany"]],
+)
+def test_clean_notebook_preserve_metadata_tags_special(
+    dirty_notebook: nbformat.NotebookNode,
+    clean_notebook_with_tags_special_metadata: nbformat.NotebookNode,
+    preserve_cell_metadata,
```
Comment (srstevenson) [inline tests/test_clean_notebook.py:86]: Ditto on type annotation.

**Review by srstevenson — CHANGES_REQUESTED** [review]
Thanks for preparing this! In general, I think the implementation is good to go. I've made a new minor comments inline.

File: README.md [hunk 8]
```diff
@@ -153,28 +160,69 @@ repos:
...
           - --remove-empty-cells
+          - --preserve-cell-metadata
+          - tags
+          - slideshow
+          - --
 ```
 
+Note that the find `--` is needed because `--preserve-cell-metadata` may take
+an arbitrary number of arguments, which need to be distinguished from the
+notebooks that `pre-commit` will append to the arguments.
```
[the remark this answers is not in the document]
Comment (tovrstra) [reply README.md:171]: Yes, that would be better. I'll also try to improve clarity of this sentence.

File: README.md [hunk 9]
```diff
@@ -153,28 +160,69 @@ repos:
...
+```
+
+To also check for empty cells, add the `-e`/`--remove-empty-cells` flag.
+To ignore cell metadata, add the `-m`/`--preserve-cell-metadata` flag,
+optionally with a selection of metadata fields to ignore.
+To ignore cell outputs, add the `-o`/`--preserve-cell-outputs` flag.
+
+`nb-clean` will exit with status code 0 if the notebook is clean, and status
+code 1 if it is not. `nb-clean` will also print details of cell execution
+counts, metadata, outputs, and empty cells it finds.
```
[the remark this answers is not in the document]
Comment (tovrstra) [reply README.md:197]: Sorry for that. Will be fixed.

File: tests/test_check_notebook.py (hunk 2 above)
Comment (tovrstra) [reply tests/test_check_notebook.py:67]: ok, no problem

**Comment by tovrstra** [comment]
@srstevenson Thanks for the review. I should have caught these, but overlooked them somehow.

I'll go through the final diff to see I can spot anything else.

**Comment by tovrstra** [comment]
All comments should be addressed, but if something else pops up, I'm happy to fix more.

**Comment by srstevenson** [comment]
Thanks @tovrstra! I merged locally to resolve the conflicts.

**Comment by srstevenson** [comment]
This has been included in [nb-clean 2.4.0](https://pypi.org/project/nb-clean/2.4.0/), which is available from PyPI.

**Comment by tovrstra** [comment]
Thanks for taking care and making the release! We'll start using it right away.

**Pull request closed without merging**